packages/jupyterlab-manager/src/manager.ts

```diff
@@ -213,14 +332,114 @@ class WidgetManager extends ManagerBase<Widget> implements IDisposable {
     return this._rendermime;
   }
 
+  /**
+   * A signal emitted when state is restored to the widget manager.
+   *
+   * #### Notes
+   * This indicates that previously-unavailable widget models might be available now.
+   */
+  get restored(): ISignal<this, void> {
+    return this._restored;
+  }
+
   register(data: IWidgetRegistryData) {
     this._registry.set(data.name, data.version, data.exports);
   }
 
+  /**
+   * Get a model
+   *
+   * #### Notes
+   * Unlike super.get_model(), this implementation always returns a promise and
+   * never returns undefined. The promise will reject if the model is not found.
+   */
+  async get_model(model_id: string): Promise<WidgetModel> {
+    const modelPromise = super.get_model(model_id);
+    if (modelPromise === undefined) {
+      throw new Error('widget model not found');
+    }
+    return modelPromise;
+  }
+
+  /**
+   * Register a widget model.
+   */
+  register_model(model_id: string, modelPromise: Promise<WidgetModel>): void {
+    super.register_model(model_id, modelPromise);
+
+    // Update the synchronous model map
+    modelPromise.then(model => {
+        this._modelsSync.set(model_id, model);
+        model.once('comm:close', () => {
+            this._modelsSync.delete(model_id);
+        });
+    });
+    this.setDirty();
+  }
+
+
+  /**
+   * Close all widgets and empty the widget state.
+   * @return Promise that resolves when the widget state is cleared.
+   */
+  async clear_state(): Promise<void> {
+    await super.clear_state();
+    this._modelsSync = new Map();
+    this.setDirty();
+  }
+
+  /**
+   * Synchronously get the state of the live widgets in the widget manager.
+   *
+   * This includes all of the live widget models, and follows the format given in
+   * the @jupyter-widgets/schema package.
+   *
+   * @param options - The options for what state to return.
+   * @returns Promise for a state dictionary
+   */
+  get_state_sync(options: IStateOptions = {}) {
+      const models = [];
+      for (let model of this._modelsSync.values()) {
+        if (model.comm_live) {
+          models.push(model);
+        }
+      }
+      return serialize_state(models, options);
+  }
+
+  /**
+   * Set the dirty state of the notebook model if applicable.
+   *
+   * TODO: should also set dirty when auto-save state is turned on (easy), and when any
+   * model changes any data (likely hard right now).
```

Comment:
Note to self: Just before merging, if this TODO note is still here, it should be replaced by an issue.